### PR TITLE
Monitor for pod activity and apply image stream annotation informatio…

### PR DESCRIPTION
**Pull request by xenserverarmy**
…n to pod

Since not all pods are created from ImageStreams, pods also need to be monitored.  When a Pod lacks a backing ImageStream there isn't any place to put the existing Annotations. In this PR the Annotations are also applied to the Pod thereby allowing risk to be assessed for running Pods. The same Annotations used by the ImageStream are used in the Pods

This has been tested with OpenShift 3.6 under the following conditions:

1. A new pod is created as a result of a pipeline build
2. A new pod is created as a result of a direct image deployment
3. A new pod is created as a result of a template execution with docker.io reference
4. Running pods are discovered upon installation and their source of truth is validated from the associated Hub instance
5. Pods are scaled up and identified and annotated
6. Builder pods are identified and annotated

### CODE REVIEW

File: common/scanresults.go [hunk 1]
```diff
@@ -215,7 +220,117 @@ func ProjectVersionResults(info ImageInfo, imageId string, taggedName string, sh
...
 	log.Printf("Found %d high severity vulnerabilities and %d policy violations for %s:%s\n", vulnerabilities, violations, taggedName, imageId[:10])
 
-	results = annotate.UpdateAnnotations(info, sha, violations, vulnerabilities, projectVersionUrl, scanId, projectVersionUI)
+	results = annotate.UpdateAnnotations(info, violations, vulnerabilities, projectVersionUrl, scanId, projectVersionUI)
 
 	return nil, results
 }
+
+// GetScanResultsFromProjectVersion obtains the vuln data from the Hub based solely on the project and version.
+func GetScanResultsFromProjectVersion(projectName string, version string, annotate *Annotator, hubConfig *HubConfig) (e error, results ImageInfo) {
```
Comment (jayunit100) [inline common/scanresults.go:229]: This is a massive function : can we break it up for testability ?

File: controller/pkg/controller/job.go [hunk 2]
```diff
@@ -23,15 +23,25 @@ under the License.
 package controller
 
 import (
+	"encoding/json"
+
 	bdscommon "github.com/blackducksoftware/ose-scanner/common"
+	osimageapi "github.com/openshift/origin/pkg/image/api"
```
Comment (jayunit100) [inline controller/pkg/controller/job.go:29]: Was assuming this commit would reduce openshift coupling .... is that happening or not yet ... just checking ?

File: controller/pkg/controller/job.go [hunk 3]
```diff
@@ -90,7 +102,70 @@ func (job Job) UpdateAnnotationInfo(newInfo bdscommon.ImageInfo) bool {
...
+	}
+	info.Labels = pod.ObjectMeta.Labels
+	if info.Labels == nil {
+		log.Printf("Pod %s has no labels - creating object.\n", podName)
+		info.Labels = make(map[string]string)
+	}
+	return info
+}
+
+// UpdatePodAnnotationInfo updates an existing pod annotations/lables with our scna results
```
Comment (jayunit100) [inline controller/pkg/controller/job.go]: Nit : scan not scna:)

File: controller/pkg/controller/controller.go [hunk 4]
```diff
@@ -245,14 +260,120 @@ func (c *Controller) getImages(done <-chan struct{}) {
...
+
+		c.processPod(&pod)
+
+	}
+
+	return
+
+}
+
+func (c *Controller) waitPodRunning(podName string, namespace string) {
```
Comment (jayunit100) [inline controller/pkg/controller/controller.go:320]: Consider importing the e2e libs like https://github.com/openshift/origin/pull/7049/files as an example or check out http://jayunit100.blogspot.com/2016/10/importing-kuberentes-e2es-as-artifact.html?m=1

**Comment by jayunit100** [comment]
@xenserverarmy sorry i think i forgot to click.

File: controller/pkg/controller/job.go (hunk 2 above)
Comment (xenserverarmy) [reply controller/pkg/controller/job.go:29]: There is no plan to decouple OpenShift. We are firmly commited to supporting OpenShift and ImageStreams

File: controller/pkg/controller/controller.go (hunk 4 above)
Comment (xenserverarmy) [reply controller/pkg/controller/controller.go:320]: Something to look at in the future, but I feel out of scope for this PR. Likely something to have as an independent feature at some point.

File: common/scanresults.go (hunk 1 above)
Comment (xenserverarmy) [reply common/scanresults.go:229]: If you look closely, its not actually that big. What's happening is a series of API calls to Hub are occuring with the results being parsed. From a testability perspective, the low hanging fruit is to have the common/hubserver.go and common/types.go automatically generated as a set of API bindings which can then be vetted. Next up would be the consumption side.

File: controller/pkg/controller/job.go (hunk 3 above)
Comment (xenserverarmy) [reply controller/pkg/controller/job.go]: Fixed

**Review by jayunit100 — APPROVED** [review]
Approved as wip

**Comment by jayunit100** [comment]
Merged , feel free to ping us in hipchat CloudNativeEngineering room, someone's always around for reviews .

**Pull request merged**